File: finagle-core-18.2.0/src/main/java/com/twitter/finagle/builder/ClientBuilder.java
package com.twitter.finagle.builder;

import com.newrelic.api.agent.weaver.MatchType;
import com.newrelic.api.agent.weaver.Weave;
import com.newrelic.api.agent.weaver.Weaver;
import com.newrelic.instrumentation.finagle.core.NewRelicServiceWrapper;
import com.twitter.finagle.Service;
import com.twitter.finagle.Stack;
import com.twitter.finagle.param.Label;

@Weave(type = MatchType.BaseClass)
public abstract class ClientBuilder<Req, Rep, HasCluster, HasCodec, HasHostConnectionLimit> {

	/*
	 * Called when a new service is constructed.  We wrap the resulting service with our wrapper service
	 */
	public Service<Req, Rep> build(com.twitter.finagle.builder.ClientConfigEvidence<HasCluster, HasCodec, HasHostConnectionLimit> c) {
		Service<Req, Rep> service = Weaver.callOriginal();
		Stack.Params params = params();
		boolean value = params.contains(Label.param());
		NewRelicServiceWrapper<Req, Rep> wrapper = new NewRelicServiceWrapper<>(service);
		if(value) {
			Label label = params.apply(Label.param());
			wrapper.setLabel(label.label());
			
		}
		return wrapper;
	}
	
	public abstract Stack.Params params();
}

File: finagle-core-18.2.0/src/main/scala/com/twitter/finagle/builder/ClientBuilder.scala
package com.twitter.finagle.builder

import com.newrelic.api.agent.weaver.scala.{ScalaMatchType, ScalaWeave}
import com.newrelic.api.agent.Trace
import com.newrelic.instrumentation.finagle.core.ServiceNames
import com.newrelic.api.agent.weaver.Weaver
import com.twitter.finagle.{Name, Service,ServiceFactory}
import com.twitter.finagle.client.StackBasedClient

@ScalaWeave(`type` = ScalaMatchType.Object, originalName="com.twitter.finagle.builder.ClientBuilderClient")
class ClientBuilderClient_instrumentation {

    def newService[Req, Rep](client0: StackBasedClient[Req, Rep],dest: Name,label: String  ): Service[Req, Rep] = {
      val value : Service[Req, Rep] = Weaver.callOriginal()
      ServiceNames.addServiceName(value,dest,label)
      value
  }

   def newClient[Req, Rep](client: StackBasedClient[Req, Rep], dest: Name, label: String): ServiceFactory[Req, Rep] = {
      val value : ServiceFactory[Req, Rep]  = Weaver.callOriginal()
      ServiceNames.addServiceFactoryName(value,dest,label)
      value
   }
}